### PR TITLE
[SPARK-19716][SQL] support by-name resolution for struct type elements in array

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Previously when we construct deserializer expression for array type, we will first cast the corresponding field to expected array type and then apply `MapObjects`.

However, by doing that, we lose the opportunity to do by-name resolution for struct type inside array type. In this PR, I introduce a `UnresolvedMapObjects` to hold the lambda function and the input array expression. Then during analysis, after the input array expression is resolved, we get the actual array element type and apply by-name resolution. Then we don't need to add `Cast` for array type when constructing the deserializer expression, as the element type is determined later at analyzer.

## How was this patch tested?

new regression test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @marmbrus @liancheng

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/ReduceAggregator.scala [hunk 1]
```diff
@@ -27,19 +27,17 @@ import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
  *
  * This class currently assumes there is at least one input row.
  */
-private[sql] class ReduceAggregator[T: Encoder](func: (T, T) => T)
+private[sql] class ReduceAggregator[T](func: (T, T) => T)(@transient implicit val enc: Encoder[T])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/ReduceAggregator.scala]: we should not serialize an unresolved encoder, so I make it `transient` here

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala [hunk 2]
```diff
@@ -62,6 +66,54 @@ class EncoderResolutionSuite extends PlanTest {
     encoder.resolveAndBind(attrs).fromRow(InternalRow(InternalRow(str, 1.toByte), 2))
   }
 
+  test("real type doesn't match encoder schema but they are compatible: array") {
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala]: Is there a reason we don't have any `encodeDecodeTest`?

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala]: `encodeDecodeTest` is a round trip test that the type is always same. I added an end-to-end test in `DatasetSuite`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 3]
```diff
@@ -581,17 +592,24 @@ case class MapObjects private(
           // collection
           val collObjectName = s"${cls.getName}$$.MODULE$$"
           val getBuilderVar = s"$collObjectName.newBuilder()"
-
-          (s"""${classOf[Builder[_, _]].getName} $builderValue = $getBuilderVar;
-        $builderValue.sizeHint($dataLength);""",
+          (
+            s"""
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: below are style-only changes

**Comment by liancheng** [comment]
LGTM. Merging to master. Thanks!

**Pull request closed without merging**